### PR TITLE
Implement Convolution Layer

### LINKED ISSUE #10: Implement convolution layer

### PR DESCRIPTION
**Pull request by AndreySorokin7**
Close #10

### CODE REVIEW

File: include/layers/ConvLayer.hpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+           const std::vector<ValueType>& kernel, size_t step_);
+};
+template <typename ValueType>
+void ConvolutionalLayer<ValueType>::run(const Tensor& input, Tensor& output,
+                                        const std::vector<ValueType>& kernel_,
+                                        size_t step_) {
+  std::vector<ValueType> matrix = *input.as<ValueType>();
+  size_t input_size = matrix.size();
+  size_t kernel_size = static_cast<int>(std::sqrt(kernel_.size()));
+  int input_width = static_cast<int>(std::sqrt(input_size / 3));
```
Comment (NeiroYT) [inline include/layers/ConvLayer.hpp:26]: This doesn't make sense if you can work with shapes, make kernel tensor, and input  have its shape

File: include/layers/ConvLayer.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  std::vector<ValueType> outputvec;
+  for (int i = input_width + static_cast<int>((kernel_size - 1) / 2);
+       i < static_cast<int>(input_size / 3); i += static_cast<int>(step_)) {
```
Comment (NeiroYT) [inline include/layers/ConvLayer.hpp:29]: Is it better to find center instead of left-top corner?

File: include/layers/ConvLayer.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      }
+      outputvec.push_back(color);
+    }
+    if ((i + static_cast<int>((kernel_size - 1) / 2) + 1) % input_width == 0) {
+      if (i + input_width + static_cast<int>((kernel_size - 1)) ==
+          static_cast<int>(input_size / 3)) {
+        i += input_width + static_cast<int>((kernel_size - 1)) + 1;
+      } else {
+        i += input_width * (static_cast<int>(step_) - 1) +
+             (3 - static_cast<int>(step_));
```
Comment (NeiroYT) [inline include/layers/ConvLayer.hpp:47]: I can't understand this without comments

**Review by NeiroYT — CHANGES_REQUESTED** [review]

File: include/layers/ConvLayer.hpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
+#pragma once
+#include <cmath>
+#include <stdexcept>
+#include <vector>
+
+#include "layers/Layer.hpp"
+
+class ConvolutionalLayer : public Layer {
+ public:
+  ConvolutionalLayer() = default;
```
Comment (aobolensk) [inline include/layers/ConvLayer.hpp:10]: Missing parameters:
- stride
- pads (horizontal, vertical)
- dilations

File: include/layers/ConvLayer.hpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                  size_t step_);
+};
+void ConvolutionalLayer::run(const Tensor& input, Tensor& output,
+                             const Tensor& kernel_, size_t step_) {
+  switch (input.get_type()) {
+    case Type::kInt: {
+      std::vector<int> matrix = *input.as<int>();
+      size_t input_size = input.get_shape()[input.get_shape().dims() - 2] *
+                          input.get_shape()[input.get_shape().dims() - 3];
+      size_t kernel_size = kernel_.get_shape()[kernel_.get_shape().dims() - 1];
```
Comment (aobolensk) [inline include/layers/ConvLayer.hpp:21]: Need to add checks if there are enough dims for that

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: include/layers/ConvLayer.hpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+ public:
+  ConvolutionalLayer() = default;
+  ConvolutionalLayer(size_t step, size_t pads, size_t dilations) {
+    stride_ = step;
+    pads_ = pads;
+    dilations_ = dilations;
+  }
+  void run(const Tensor& input, Tensor& output, const Tensor& kernel_);
+};
+void ConvolutionalLayer::run(const Tensor& input, Tensor& output,
```
Comment (NeiroYT) [inline include/layers/ConvLayer.hpp:95]: you can move run function to cpp file

File: include/layers/ConvLayer.hpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
+#pragma once
+#include <cmath>
+#include <stdexcept>
+#include <vector>
+
+#include "layers/Layer.hpp"
+
+template <typename T>
+void emplConv(const Tensor& input, Tensor& output, const Tensor& kernel_,
```
Comment (NeiroYT) [inline include/layers/ConvLayer.hpp:9]: `ConvImpl`, implementation

**Review by allnes — APPROVED** [review]

File: include/layers/ConvLayer.hpp (hunk 7 above)
Comment (aobolensk) [reply include/layers/ConvLayer.hpp:9]: Please, follow up this

**Review by aobolensk — APPROVED** [review]

**Review by NeiroYT — APPROVED** [review]
4 tests are not enough i think, but i guess it can be fixed later with parameterized tests

**Comment by aobolensk** [comment]
> 4 tests are not enough i think, but i guess it can be fixed later with parameterized tests

I agree here

**Pull request merged**